File: server/storage-service/src/main/java/com/mshop/storageservice/controller/ImageController.java
package com.mshop.storageservice.controller;

import com.sudo248.domain.base.BaseResponse;
import com.mshop.storageservice.config.StorageSource;
import com.mshop.storageservice.service.ImageService;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@RestController
@RequestMapping("/images")
public class ImageController {

    private final ImageService imageService;

    public ImageController(ImageService imageService) {
        this.imageService = imageService;
    }

    @PostMapping(value = "/upload", consumes = {MediaType.MULTIPART_FORM_DATA_VALUE})
    public ResponseEntity<BaseResponse<?>> upload(
            @RequestParam(value = "image") MultipartFile image,
            @RequestParam(value = "source", required = false, defaultValue = "cloudinary") String source
    ) {
        return imageService.storeImage(image, StorageSource.fromValue(source));
    }
}

File: server/product-service/src/main/kotlin/com/mshop/productservice/controller/ImageController.kt
package com.mshop.productservice.controller

import com.mshop.base.BaseController
import com.mshop.base.BaseResponse
import com.mshop.productservice.dto.ImageDto
import com.mshop.productservice.service.ImageService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/discovery/images")
class ImageController(
    private val imageService: ImageService
) : com.mshop.base.BaseController() {

    @PostMapping
    suspend fun upsertImage(
        @RequestBody imageDto: ImageDto
    ): ResponseEntity<com.mshop.base.BaseResponse<*>> = handle {
        imageService.upsertImage(imageDto)
    }

    @DeleteMapping("/{imageId}")
    suspend fun deleteImage(
        @PathVariable("imageId") imageId: String
    ): ResponseEntity<com.mshop.base.BaseResponse<*>> = handle {
        imageService.deleteImage(imageId)
    }
}